pubsub/pubsub.go

```diff
@@ -182,41 +184,50 @@ type subTopic struct {
 	name   string
 }
 
-func (t *subTopic) Subscribe(ctx context.Context, handler bps.Handler, _ ...bps.SubOption) error {
+func (t *subTopic) Subscribe(handler bps.Handler, _ ...bps.SubOption) (bps.Subscription, error) {
 	// usual way to "unsubscribe" from Google PubSub is to cancel context:
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	ctx, cancel := context.WithCancel(context.Background())
 
-	sub, err := t.client.CreateSubscription(ctx, bps.GenClientID(), native.SubscriptionConfig{
+	gsub, err := t.client.CreateSubscription(ctx, bps.GenClientID(), native.SubscriptionConfig{
 		Topic: t.client.Topic(t.name),
 	})
 	if err != nil {
-		return err
+		cancel()
+		return nil, err
 	}
-	defer func() {
-		// give subscription 5s to delete:
-		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-		defer cancel()
 
-		_ = sub.Delete(ctx)
-	}()
+	sub := concurrent.NewGroup(ctx)
 
-	// sub.Receive calls handler concurrently from multiple goroutines, so synchronise it:
-	handler = &syncHandler{Handler: handler}
+	sub.Go(func() error {
```

Comment:
again, I am not quite happy with this, if any of the `Receive` call fail, we will just exit and silently stop consuming

Reply:
looks like nats has as a separate callback for that. we need to add that too (next PR) but either way, I would prefer us to retry and not to exit and just stop.

something like (not this PR, just mentioning):

```go
sub, err := sc.Subscribe(..., WithCustomErrorHandler())
...
go func() {
  for err := range sub.Errors() {
    log.Println("custom logger", err)
  }
}()
```

Reply:
I'll make a for-loop for it with, let's say, 3 retries + a TODO for custom err handler.

Reply:
Actually, I'll do nothing for it now and will only leave TODO to add WithErrorHandler, because Receive will return only when it is not retryable:

https://pkg.go.dev/cloud.google.com/go/pubsub?tab=doc#Subscription.Receive
>  Receive calls f with the outstanding messages from the subscription. It blocks until ctx is done, or the service returns a non-retryable error.

Reply:
so the idea is just not to exit, it should keep going in a for loop until someone calls `sub.Stop()`. if an error occurs, it should just print it via `log.Println()` for now. later we can add custom error handling via `Errors()` channel